src/rt/array/construction.d

```diff
@@ -0,0 +1,165 @@
+/**
+ This module contains compiler support for constructing dynamic arrays
+
+  Copyright: Copyright Digital Mars 2000 - 2019.
+  License: Distributed under the
+       $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+     (See accompanying file LICENSE)
+  Source: $(DRUNTIMESRC rt/_array/_construction.d)
+*/
+module rt.array.construction;
+
+/**
+ * Does array initialization (not assignment) from another array of the same element type.
+ * Params:
+ *  to = what array to initialize
+ *  from = what data the array should be initialized with
+ * Returns:
+ *  The constructed `to`
+ * Bugs:
+ *  The functions `@safe` safety is faked.
+ */
+T[] _d_arrayctor(T)(return scope T[] to, scope const T[] from) @trusted
+{
+    import core.internal.traits : Unqual;
+    import core.stdc.string : memcpy;
+
+    // Force `enforceRawArraysConformable` to be `pure`
+    void enforceRawArraysConformable(const char[] action, in size_t elementSize, const void[] a1, const void[] a2, in bool allowOverlap = false)
+    {
+        import rt.util.array : enforceRawArraysConformable;
+        mixin("alias Type = " ~ typeof(&enforceRawArraysConformable).stringof ~ " pure;");
+        (cast(Type)&enforceRawArraysConformable)(action, elementSize, a1, a2, allowOverlap);
+    }
+
+    debug(PRINTF) printf("_d_arrayctor(to = %p,%d, from = %p,%d) size = %d\n", from.ptr, from.length, to.ptr, to.length, T.tsize);
+
+    auto element_size = T.sizeof;
+
+    void[] vFrom = (cast(void*)from.ptr)[0..from.length];
+    void[] vTo = (cast(void*)to.ptr)[0..to.length];
+    enforceRawArraysConformable("initialization", element_size, vFrom, vTo, false);
+
+    size_t i;
+    try
+    {
+        for (i = 0; i < to.length; i++)
+        {
+            auto elem = cast(Unqual!T*)&to[i];
+            // Copy construction is defined as bit copy followed by postblit.
+            memcpy(elem, &from[i], element_size);
+
+            static if (__traits(hasMember, T, "__xpostblit") &&
+                // Bugzilla 14746: Check that it's the exact member of S.
+                __traits(isSame, T, __traits(parent, elem.__xpostblit)))
+                elem.__xpostblit();
+        }
+    }
+    catch (Exception o)
+    {
```

Comment:
Just for reference, this means calling destructors will not be done if the postblit threw a non-exception throwable. IIRC the language spec allows this behaviour.

Reply:
See https://github.com/dlang/druntime/blob/ab890e0fb09cb6142806bf3d2cfd0a363936a298/src/rt/arrayassign.d#L180-L199 for the genesis of this code.